lib/AST/DiagnosticEngine.cpp

```diff
@@ -1024,12 +1045,22 @@ void DiagnosticEngine::emitDiagnostic(const Diagnostic &diagnostic) {
     
     SmallVector<std::string, 1> educationalNotePaths;
 
-    auto associatedNotes = educationalNotes[(uint32_t)diagnostic.getID()];
-    while (associatedNotes && *associatedNotes) {
+    auto associatedNote = educationalNotes[(uint32_t)diagnostic.getID()];
+
+    if (associatedNote) {
       SmallString<128> notePath(getDiagnosticDocumentationPath());
-      llvm::sys::path::append(notePath, *associatedNotes);
-      educationalNotePaths.push_back(notePath.str().str());
-      ++associatedNotes;
+      if (!locale.empty() && locale != "en" &&
+          haveTranslationForEducationalNote(diagnostic.getID(), locale,
+                                            educationalNoteTranslations)) {
```

Comment:
I think we should associate education notes with `Diagnostic` and let `LocalizationProducer` implementations handle them just like regular diagnostic strings, WDYT?

Reply:
yeah, I think that makes the most sense long-term. I'm going to hold off on merging this PR for now until I've had a chance to look into that more in case it significantly changes the approach.

Reply:
Sounds good!